providers/extensible_schema.go

```diff
@@ -105,13 +117,45 @@ func (x *extensibleSchema) LookupField(resource string, field string) (*resource
 		return found, nil
 	}
 
-	x.loadAllSchemas()
-	x.refresh()
+	x.unsafeLoadAll()
+	x.unsafeRefresh()
 
 	return found, found.Fields[field]
 }
 
-func (x *extensibleSchema) Add(name string, schema *resources.Schema) {
+// ---------------------------- unsafe methods ----------------------------
+// |  Only use these calls inside of a lock.                              |
+// |  Do NOT lock the object during these calls.                          |
+// V  Do NOT call to locking methods (~ everything above this line).      V
+// ------------------------------------------------------------------------
+
+func (x *extensibleSchema) unsafeLoadAll() {
+	// If another goroutine started to load this before us, it will be locked until
+	// we complete to load everything and then it will be dumped into this
+	// position. At this point, if it has been loaded we can return safely, since
+	// we don't unlock until we are finished loading.
+	if x.lastRefreshed >= LastProviderInstall {
+		return
+	}
+	x.lastRefreshed = LastProviderInstall
+
+	providers, err := ListActive()
+	if err != nil {
+		log.Error().Err(err).Msg("failed to list all providers, can't load additional schemas")
+		return
+	}
+
+	for name := range providers {
+		schema, err := x.runtime.coordinator.LoadSchema(name)
```

Comment:
Is it possible for `LoadSchema` to get called concurrently. It looks to me like different runtimes share coordinators. `LoadSchema` has a path where it can end up modifying the data the coordinator holds through `provider.LoadResources()`.

Reply:
+1, yes, I think we need to separately make the coordinator thread-safe.